share/doc/revng/references/abi-definition.md

```diff
@@ -0,0 +1,213 @@
+
+Let's look at an ABI that allocates four registers for passing function arguments (`r0-r3`). Then, for a function like
+```cpp
+void function(uint32_t, Big);
+```
+the `uint32_t` argument would be passed in the first GPR (`r0`).
+
+But that would also mean that the remaining three available GPRs are not enough to fit the entirety of the `Big` object, meaning it needs to either be split between the registers and the memory, or passed using the stack. That's exactly what this option states.
+
+## `BigArgumentsUsePointersToCopy` (`bool`)
+
+This allows a non-position-based ABI to take advantage of one of the main traits of position-based ABI: ability to use pointers to copy to pass big aggregates.
+
+This is one of the defining features of arm64 ABI, which allows it to get the best of both approaches - an ability to pass aggregates by copy in registers from the non-position-based approach while also being able to sometimes avoid a copy of a bigger aggregate from the position-based approach.
+
+## `NoRegisterArgumentsCanComeAfterStackOnes` (`bool`)
+
+States whether ABI allows a stack argument (mainly if it's too big to be placed in the registers) to precede other register arguments.
+
+For example, if there is a `Big` struct that has to use the stack, and
+a function like
+```cpp
+void function(Big, uint32_t);
+```
+if this value is set to true, both argument will be passed on stack, otherwise, only the struct will.
+
+## `AllowUnnaturallyAlignedTypesInRegisters` (`bool`)
+
+States whether ABI allows unnaturally aligned types[^1] to be passed in registers. When set to false, all the unnatural types will only be put on stack.
+
+## `CalleeIsResponsibleForStackCleanup` (`bool`)
+
+Specifies who is responsible for cleaning the stack after the function call. If equal to `true`, it's the callee, otherwise it the caller.
+
+## `FloatsUseGPRs` (`bool`)
+
+Setting this to true disables usage of vector registers entirely.
+
+## `StackArgumentsUseRegularStructAlignmentRules` (`bool`)
+
+Setting this to true allows stack arguments to use slots smaller than a word (makes them behave like a single struct instead of a bunch of arguments): so they don't get extended unnecessarily. A notable ABI using this is apple flavor of the aarch64.
+
+For example
+```cpp
+void function(/* enough arguments of all the registers */, uint8_t, uint8_t);
+```
+When this is set to true, both `uint8_t` will get placed as if they were a struct, that is without any padding in-between. When this is false, both of them will get extended to register size before getting placed, leading to padding.
+
+## `UseStrictAggregateAlignmentRules` (`bool`)
+
+Treat all the structs passed using the stack as if they were aligned unnaturally[^1].
+
+This means that they are going to handled with extreme care (which usually just means extra padding).
```

Comment:
Obscure, I don't get what this means.

Reply:
When about something like
```
This means that options that assume regular alignment rules (like `StackArgumentsUseRegularStructAlignmentRules`) stop applying to them, while others (such as `AllowUnnaturallyAlignedTypesInRegisters`) start.
```
?

Reply:
This is so hard to get right. I feel like the default behavior you're talking about is somewhat implicit, and beside the point of this option: it's about which _other options_ should apply to 
this type, not about the alignment specifics.

Reply:
Either way, I added an extra note:

> Practically this option is mostly used to conditionally ensure [`StackArgumentsUseRegularStructAlignmentRules`](#stackargumentsuseregularstructalignmentrules-bool) does not apply to structs (see apple aarch64 abifor reference).

Do you still feel like this is not enough?

Reply:
> > It would be nice to have some section describing how naturally aligned structs (and maybe unnaturally aligned structs) are "normally" handled and then this option can read more easily as an exception to what's normal.
>
> the default behavior you're talking about is somewhat implicit, and beside the point of this option

What I meant here is a section on the top of the document, not in the description of this specific option.

I'm still confused about this set of options:
1. Maybe we should replace the term "unnaturally aligned" with "packed" all over the place. It's much more intuitive to me.
2. About `StackArgumentsUseRegularStructAlignmentRules`: can we negate this option and rename this to `PackStackArguments`? I feel like "stack arguments use regular struct alignment rules" states what's normal while here we want to specify an exceptional behavior. Alternatively we could call this `AlignStackArguments`, but I feel `PackStackArguments"`makes you immediately think "no padding", which is what happens in practice.
3. `UseStrictAggregateAlignmentRules` description mentions that `AllowUnnaturallyAlignedTypesInRegisters` "starts applying", which applies to register arguments, but `UseStrictAggregateAlignmentRules` seems to apply to stack arguments only. I'm confused.
   Does `UseStrictAggregateAlignmentRules` apply to register arguments? If not, don't mention `AllowUnnaturallyAlignedTypesInRegisters`.
4. AFAIU, `StackArgumentsUseRegularStructAlignmentRules` means "pack together stack arguments".
   `UseStrictAggregateAlignmentRules` means "behave as if the struct passed on the stack were packed".
   Is this correct?
   We need to document somewhere (at the top?) what's the "standard" behavior for non-packed (naturally aligned) and packed (unnaturally aligned) structs on the stack. We can't customize something that has not been defined.

Reply:
1. A single struct can be both packed and naturally aligned, they are not exclusive.
2. Technically we could, but reversing it would mean using the non-default value for every ABI but one, which is sad. Maybe we should just leave a TODO and do that once "default value" in TTG is implemented?
2.1. That being said, I do agree that `PackStackArguments` fits it better! Thanks for the idea!
3. `UseStrictAggregateAlignmentRules` (was named something along the lines of `treat all structs as if they were unnaturally aligned`), hence why all the "unnatural alignment facing" options start to apply. That being said, I see why it would be confusing for those who are not aware of that history. Should we rename it again?
4. I'll see what I can do!

Reply:
2. Actually, on the second thought

> I feel like "stack arguments use regular struct alignment rules" states what's normal while here we want to specify an exceptional behavior.

What is says is that ABI uses the normal behavior for _struct fields_ instead of the normal behavior for _stack arguments_. But yeah, good catch, word `pack` fits it really well.

Reply:
1. If a `struct` is packed but happens to be naturally aligned, I would define it as "not packed" anyway. We can give such definition at the top.
   I understand that this might be confusing, since we unconditionally emit all of our structs as `packed`, but if a C programmer uses `packed`, it's because he wants them to ignore alignment rules, otherwise it would just leave them out.
   I find "packed" so much more intuitive than "unnaturaly aligned" (which makes me think of some mutatant monster).
   One day we might even decide to remove `packed` from some of our `struct`s in the decompiled C code, more readable at the cost of less portable C, even if in the end the C code we emit is already tied to an ABI. But let's stay focused on the original point.
2. > using the non-default value for every ABI but one

   That's fine with me.
3. Let's ignore history entirely and find a suitable name. What do you propose?
4. :+1:

Reply:
1. I still have some doubts, but if you're believe in it so strongly, sure.

3. The point of the option is kind of straightforward: "hey, struct handling in this ABI sucks, treat _them all_ as carefully as you possibly can". But expressing is shorter than that is kind of hard. And, even after your argument for packed in `1` which I can kind of get behind, it doesn't really hold in here, since all of our structs are _packed_, that would probably just end up confusing people. And if we are dropping `unnaturally-aligned` thing elsewhere, it's definitely not worth keeping it only here - for consistency sake if for nothing else. And those two were my best bets 😅.

I'm starting to doubt whether dropping `unnaturally-aligned` term is a good idea 🤔.

Reply:
2. 1. I've absolutely no idea what "treating arguments *carefully*" means... Like, are we being sloppy with others? You have this concept of "careful" (I've seen such comments in the code too) which don't bind to anything for me. It seems like you've been working so much on this that certain things are obvious to you, but they are not obvious to others.
   2. As for the confusion on the term `packed`, I really think we should stop emit all structs as packed, or at least have an option to emit them non-`packed` if the chosen ABI for the backend allows that.
   Apart from this, as long as we put a warning at the top, I think we're fine.
   Consider `StackArgumentsUseRegularStructAlignmentRules`, would you prefer to hear "you have to pretend those two `uint8_t` are in a `struct` and that `struct` respects the standard alignment conventions (which you don't remember)" or just "those two `uint8_t` are packed together".

Reply:
> It seems like you've been working so much on this that certain things are obvious to you, but they are not obvious to others.

I wouldn't be surprised to be honest.

That being said, by "carefully", I mean taking all the precautions: as in full alignment everywhere around it (padding, yay) as well as not using registers (aside from some ABIs, like 32-bit, ARM that specifically allow it) for any it.

> an option to emit them non-`packed` if the chosen ABI for the backend allows that.

Honestly, I wouldn't mind adding a new `StructABI` field to the binary and then, as long every field of said struct makes sense in terms of the abi (we already do very similar checks when converting stack struct into separate arguments, shouldn't be hard to extrapolate it out).

At the very least we should do
```cpp
for (auto Field : Struct.Fields())
  if (Field.Offset() % ABI.alignment(Field.Type()) == 0 && ABI.isNaturallyAligned(Field.Type()))
    assume_struct_is_not_packed();
```
but I'd go full route of "deconstructing" and then "reconstructing" struct just to be sure. Since we rarely get more than handful of fields, it shouldn't be computationally expensive and I would be able to sleep better at night :D.

Going to those length would let us enable "non-packed struct" by default without many downsides. And it shouldn't be hard either, since, again, all the work is already done, it's the question of factoring the relevant piece of code out and making it callable from outside.

But, yeah, if you are on board with something like this, it should be its own PR.

> Consider `StackArgumentsUseRegularStructAlignmentRules`, would you prefer to hear "you have to pretend those two `uint8_t` are in a `struct` and that `struct` respects the standard alignment conventions (which you don't remember)" or just "those two `uint8_t` are packed together".

Honestly, agreed. I think the reason I preferred that way of phrasing it is because of sensible defaults: I'd rather it inherited the struct behavior from the same ABI in case we ever introduce a new one, rather than trying to define an extra concept of "packing" which ABI currently is not aware of.

That being said, in the hindsight, I was definitely overthinking it. Let's go with packed terminology ~~for now~~ until we have a reason not to.

Reply:
> At the very least we should do

Actually, now that I think about it, we might be able to get away with just `ABI.isNaturallyAligned(Struct)` 🤔. But that would need extensive testing.

Reply:
:+1: